packages/react-instantsearch-hooks-dom/src/ui/Breadcrumb.tsx

```diff
@@ -47,9 +47,12 @@ type UseBreadcrumbRenderState = ReturnType<typeof useBreadcrumb>;
 
 type BreadcrumbItem = UseBreadcrumbRenderState['items'][number];
 
-export type BreadcrumbProps = React.HTMLAttributes<HTMLDivElement> &
+export type BreadcrumbProps = React.ComponentProps<'div'> & {
+  classNames?: Partial<BreadcrumbClassNames>;
+};
+
+export type BreadcrumbInternalProps = BreadcrumbProps &
```

Comment:
should these two names be swapped? the ui component props is everything, but I find it hard to find a name for them. I still think that everything which is not required should be allowed, like 

```ts
type RequiredKeys<T> = { [K in keyof T]-?: {} extends Pick<T, K> ? never : K }[keyof T];
type OptionalKeys<T> = { [K in keyof T]-?: {} extends Pick<T, K> ? K : never }[keyof T];

// in widget:
export type BreadcrumbProps = OptionalKeys<BreadcrumbUiComponentProps> & UseBreadcrumbProps
```

This way what we'll need to be careful for is that the optionality of the keys is correct, so separator here for example should be required, but have `string | undefined` as its value

What do you think?

Reply:
To make sure I understood, we would keep all props in a single type, in which we have:
- required props → specific to the ui component
- optional props → passed to extend the widget props

To illustrate, the type for `<Breadcrumb>` would be: 
```ts
type BreadcrumbProps = React.ComponentProps<'div'> &
  Required<Pick<UseBreadcrumbRenderState, 'items' | 'createURL'>> & {
    classNames?: Partial<BreadcrumbClassNames>;
    hasItems: boolean;
    onNavigate: UseBreadcrumbRenderState['refine'];
    separator: string | undefined;
    translations: BreadcrumbTranslations;
  };
```

If this is what you had in mind, I like the idea, but there is a risk of forgetting this custom convention in the future, because it is not really self-explanatory.

Reply:
An alternative could be:

```tsx
type WrapperProps = React.ComponentProps<'div'>;
type RequiredProps = Required<
  Pick<UseBreadcrumbRenderState, 'items' | 'createURL'> & {
    hasItems: boolean;
    onNavigate: UseBreadcrumbRenderState['refine'];
    separator: string | undefined;
    translations: BreadcrumbTranslations;
  }
>;
type OptionalProps = Partial<{
  classNames: Partial<BreadcrumbClassNames>;
}>;

export type BreadcrumbProps = WrapperProps & RequiredProps & OptionalProps;
```


This puts a little more structure, and less likely to have it wrong. What do you think? 


The part of that I'm not sure of is using `UseBreadcrumbProps. Once ui components is standalone later, it won't have access to the React InstantSearch props, so it might be better to retype it there. But we don't have to make that last decision now.

Reply:
Thanks for the suggestion!

Reply:
This structure necessited to change the signature of some components that were previously reviewed. Specifically, props that would not be passed on the widget but were optional and defaulted, are now required. I think this is fine since we don't expose this. An example of this type of change can be seen in [ui/CurrentRefinements.tsx](https://github.com/algolia/react-instantsearch/pull/3440/files#diff-fe8163ba90a2676bee0ac9f1d8754ea8654d0776783d0212a0b81004498ab55d).

I also added a `WidgetProps<UiComponentProps>` utility type to make props forwarding easier.